python/tvm/autotvm/task/relay_integration.py

```diff
@@ -0,0 +1,331 @@
+# pylint: disable=unused-variable,invalid-name
+"""
+Decorator and utilities for the integration with TOPI and Relay
+99.9% copy-paste of implementation by @MerryMercy
+
+"""
+import warnings
+import logging
+
+
+from ... import tensor, placeholder, create_schedule, target as _target
+
+from ..util import get_const_tuple
+from .task import create, register
+
+logger = logging.getLogger('autotvm')
+
+def serialize_args(args):
+    """serialize arguments of a topi function to a hashable tuple.
+
+    Parameters
+    ----------
+    args: list of hashable or Tensor
+    """
+    ret = []
+    for t in args:
+        if isinstance(t, tensor.Tensor):
+            ret.append(('TENSOR', get_const_tuple(t.shape), t.dtype))
+        else:
+            ret.append(t)
+    return tuple(ret)
+
+
+def deserialize_args(args):
+    """The inverse function of :code:`serialize_args`.
+
+    Parameters
+    ----------
+    args: list of hashable or Tensor
+    """
+    ret = []
+    for t in args:
+        if isinstance(t, tuple) and t[0] == 'TENSOR':
+            ret.append(placeholder(shape=t[1], dtype=t[2]))
+        else:
+            ret.append(t)
+    return ret
+
+
+# Task extractor for relay program
+class TaskExtractEnv:
+    """Global environment for extracting tuning tasks from relay program"""
+    current = None
+
+    def __init__(self):
+        import topi
+        from tvm import relay
+
+        # NOTE: To add more ops, you only need to change the following lists
+        # relay op -> topi compute
+        self.op2topi = {
+            relay.op.nn.conv2d: [topi.nn.conv2d, topi.nn.depthwise_conv2d_nchw,
```

Comment:
This `op2topi` is the only different part between `TaskExtractEnv` for nnvm and relay.
I think we can move `TaskExtractEnv` to topi_integration.py, and let nnvm and relay part pass in this argument.

Reply:
The signatures are also a little different because nnvm and relay require different parameters (e.g., relay gets param dict, nnvm requires shape). How should we handle this?

Reply:
I am referring to the class `TaskExtractEnv`, which is almost TOPI level.

Reply:
Ok, good call, we can merge these.

Reply:
One thing to notice is that we pass params to relay.build, which means we will do constant folding and call tvm functions.
Then the python multiprocessing package will became in valid and we cannot tune with multiprocessing. Do you notice this in your tuning?  Possbile workaroud is to run `relay.build` in a separate thread.

Reply:
Do you mean that the tuning will crash? I didn't notice this yet; but so far I have only tried with x86.

Reply:
Will only use one thread and be very slow

Reply:
ok, do we just need to wrap the build call in another thread? That is straightforward to add.

Reply:
Yes.

Reply:
The CI breaks now after merging in the test when nnvm is imported. Is this a circular import?

Reply:
I think so. Because this is a tvm test. This's why I put the original test to nnvm folder.
Can we

Reply:
Ok, the reason I put it there was because it was a "relay/autotvm" test which doesn't really have anything to do with nnvm. I can move it to the nnvm directory.

Reply:
No, we should not move it to nnvm. See my comment below